ibtk/include/ibtk/snapshot_utilities.h

```diff
@@ -0,0 +1,94 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2022 - 2022 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBTK_snapshot_utilities
+#define included_IBTK_snapshot_utilities
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibtk/config.h>
+
+#include <ibtk/SnapshotCache.h>
+
+#include <tbox/Pointer.h>
+
+#include <HierarchyDataOpsReal.h>
+#include <PatchHierarchy.h>
+
+#include <string>
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBTK
+{
+/*!
+ * Update a snapshot at the given time. This function copies the data stored in u_idx into the snapshot index in the
+ * cache object. The snapshotted hierarchy is also updated to the current hierarchy.
+ *
+ * This outputs an error if a snapshot with that time point within the specified tolerance can not be found.
+ *
+ * Calls to this function may use a different patch index than the one used in setSnapshot(), but the underlying
+ * data layout must be consistent. This data layout is specified by the variable used to construct the cache.
+ */
+void update_snapshot(const SnapshotCache& cache,
```

Comment:
Now that we updated the documentation I actually understand what is going on.

This function modifies the input `SnapshotCache`, right? If so, even though this works (since we modify things it owns via pointers) we should not mark it as `const` since the object is being modified. The same change should be made every time we are modifying the snapshot cache.

Reply:
What about the functions in the class itself? The iterators return a pointer to the patch hierarchy, which could in theory change. Should those not be marked as `const`? Also the getter function that returns the patch index used by the class. Should that be marked as `const`?

Reply:
I suspect the answer is "don't use const with SAMRAI" since doing things by pointers evades, for language reasons, constness checks. Honestly I'd probably just remove const everywhere when dealing with `tbox::Pointer`s.

Reply:
In that case, should we change all these functions to take `shared_ptr`s instead (or maybe `tbox::Pointer` and switch to shared whenever we go to SAMRAI 4)?

Reply:
It probably isn't worth it. As soon as pointers get involved its just too hard to maintain const-correctness - that's a problem with any PatchHierarchy, regardless of how we get it (const or not).

Reply:
I don't think I was clear. I mean instead of references to `SnapshotCache`, should these external functions take `std::shared_ptr<SnapshotCache>` as arguments?

Reply:
I can also remove `const` from the class definitions.

Reply:
No - I don't see why we should limit ourselves to shared pointers managing `SnapshotCache`s. Is there some advantage to this?

Reply:
No. It's just what SAMRAI 4 does.

Reply:
Fixed